lib/src/editor.dart

```diff
@@ -0,0 +1,499 @@
+  /// ```dart
+  /// final doc = YamlEditor("YAML: YAML Ain't Markup Language");
+  /// final node = doc.parseAt(['YAML']);
+  ///
+  /// print(node.value); // Expected output: "YAML Ain't Markup Language"
+  ///
+  /// doc.assign(['YAML'], 'YAML');
+  ///
+  /// final newNode = doc.parseAt(['YAML']);
+  ///
+  /// // Note that the value does not change
+  /// print(newNode.value); // "YAML"
+  /// print(node.value); // "YAML Ain't Markup Language"
+  /// ```
+  ///
+  /// Ensuring that you get the value you want:
+  /// ```dart
+  /// final doc = YamlEditor('{a: {d: 4}, c: ~}');
+  /// doc.parseAt(['b', 'd']); // ArgumentError
+  /// doc.parseAt(['b']); // ArgumentError
+  /// doc.parseAt(['b'], orElse: null); // YamlScalar(null)
+  /// doc.parseAt(['b'], orElse: {"a": 42}); // YamlMap({"a": 42})
+  /// doc.parseAt(['b'], orElse: 42); // YamlScalar(42)
+  ///
+  /// final doc2 = YamlEditor('[0,1]');
+  /// doc2.parseAt([2]); // ArgumentError
+  /// doc2.parseAt(["2"]); // ArgumentError
+  /// ```
+  ///
+  /// [orElse] can also be used to determine if a path exists.
+  /// ```dart
+  /// doc.parseAt(my_path, orElse: #doesNotExist);
+  ///
+  /// // Or, if you know that your collection will not contain null,
+  /// doc.parseAt(my_path, orElse: null);
+  /// ```
+  @override
+  YamlNode parseAt(Iterable<Object> path, {Object orElse = #noArg}) {
+    try {
+      return _traverse(path);
+    } on ArgumentError {
+      if (orElse == #noArg) {
+        rethrow;
+      } else {
+        if (orElse is YamlNode) {
+          return orElse;
+        }
+
+        return yamlNodeFrom(orElse);
+      }
+    }
+  }
+
+  /// Sets [value] in the [path].
+  ///
+  /// Note that [assign] provides a different result as compared to a [remove] followed by an
+  /// [insertIntoList], because it preserves comments at the same level.
+  ///
+  /// Throws an [ArgumentError] if path is invalid, and throws the usual Dart errors otherwise (e.g.
+  /// [RangeError] if and index is negative or longer than list length).
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('''
+  ///   - 0
+  ///   - 1 # comment
+  ///   - 2
+  /// ''');
+  /// doc.assign([1], 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// ```yaml
+  ///   - 0
+  ///   - test # comment
+  ///   - 2
+  /// ```
+  ///
+  /// ```dart
+  /// final doc2 = YamlEditor('''
+  ///   - 0
+  ///   - 1 # comment
+  ///   - 2
+  /// ''');
+  /// doc2.remove([1]);
+  /// doc2.insertIntoList([], 1, 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// ```yaml
+  ///   - 0
+  ///   - test
+  ///   - 2
+  /// ```
+  @override
+  void assign(Iterable<Object> path, Object value) {
+    if (path.isEmpty) {
+      final end = getContentSensitiveEnd(_contents);
+      final edit = SourceEdit(0, end, getFlowString(value));
+
+      _performEdit(edit, path, yamlNodeFrom(value));
+      return;
+    }
+
+    final collectionPath = path.take(path.length - 1);
+    final keyOrIndex = path.last;
```

Comment:
Hmm, if `path` is truly an iterable (like a generator), then we probably can't do `.length` and then read more :D

Maybe it's safer to do:
```suggestion
    final collectionPath = path.toList();
    final keyOrIndex = collectionPath.removeLast();
```

Reply:
I'll probably do something like this instead!

```dart
    final pathAsList = path.toList();
    final collectionPath = pathAsList.take(path.length - 1);
    final keyOrIndex = pathAsList.last;
```

haha just for my convenience because I feel like there will be a day when I read it and forget that `removeLast()` will affect `collectionPath` as well and be really confused